libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumentationServiceImpl.java

```diff
@@ -53,94 +55,91 @@ public Map<MethodKey, CheckerMethod> lookupMethodsToInstrument(String entitlemen
-                        if (descriptor.equals(instrumentationTargetAnnotationDescriptor)) {
-                            return new InstrumentationTargetAnnotationVisitor(
-                                av,
-                                className,
-                                methodName,
-                                methodDescriptor,
-                                methodsToInstrument
-                            );
-                        }
-                        return av;
-                    }
-                };
+            public MethodVisitor visitMethod(
+                int access,
+                String checkerMethodName,
+                String checkerMethodDescriptor,
+                String signature,
+                String[] exceptions
+            ) {
+                var mv = super.visitMethod(access, checkerMethodName, checkerMethodDescriptor, signature, exceptions);
+
+                var checkerMethodArgumentTypes = Type.getArgumentTypes(checkerMethodDescriptor);
+                var methodToInstrument = parseCheckerMethodSignature(checkerMethodName, checkerMethodArgumentTypes);
+
+                var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
+                var checkerMethod = new CheckerMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
+
+                methodsToInstrument.put(methodToInstrument, checkerMethod);
+
+                return mv;
             }
         };
         reader.accept(visitor, 0);
         return methodsToInstrument;
     }
 
-    private static class InstrumentationTargetAnnotationVisitor extends AnnotationVisitor {
-        private final String checkerClassName;
-        private final String checkerMethodName;
-        private final String checkerMethodDescriptor;
-        private final HashMap<MethodKey, CheckerMethod> methodsToInstrument;
-        private String targetClassName;
-        private String targetMethodName;
-        private boolean targetMethodIsStatic;
-
-        InstrumentationTargetAnnotationVisitor(
-            AnnotationVisitor av,
-            String checkerClassName,
-            String checkerMethodName,
-            String checkerMethodDescriptor,
-            HashMap<MethodKey, CheckerMethod> methodsToInstrument
-        ) {
-            super(Opcodes.ASM9, av);
-            this.checkerClassName = checkerClassName;
-            this.checkerMethodName = checkerMethodName;
-            this.checkerMethodDescriptor = checkerMethodDescriptor;
-            this.methodsToInstrument = methodsToInstrument;
+    private static final Type CLASS_TYPE = Type.getType(Class.class);
+
+    static MethodKey parseCheckerMethodSignature(String checkerMethodName, Type[] checkerMethodArgumentTypes) {
+
+        var methodNameTokens = checkerMethodName.split("\\$");
+        if (methodNameTokens.length < 3 || "check".equals(methodNameTokens[0]) == false) {
```

Comment:
This approach doesn't really match how we describe the convention informally. It seems confusing to split the string at `$` characters when some of those are not actually separators, and then re-join those strings later to undo the damage.

It seems more straightforward to do it this way:
- Check that the name starts with `check$`. Throw if not.
- Compute `lastIndexOf('$')`. (Throw if it's 5.)
  - Everything before that `$` is the class name (after replacing `_` with `/`)
  - Everything after that `$` is the method name

Reply:
I think they are probably equivalent, but yes, looking for char indexes will be surely faster :) 
I'm convinced, I'll change it.